src/server/pkg/dlock/dlock.go

```diff
@@ -0,0 +1,75 @@
+// Package dlock implements a distributed lock on top of etcd.
+package dlock
+
+import (
+	"context"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+)
+
+// DLock is a handle to a distributed lock.
+type DLock interface {
+	// Lock acquries the distributed lock, blocking if necessary.  If
+	// the lock is acquired, it returns a context that should be used
+	// in any subsequent blocking requests, so that if you lose the lock,
+	// the requests get cancelled correctly.
+	Lock(context.Context) (context.Context, error)
+	// Unlock releases the distributed lock.
+	Unlock() error
```

Comment:
This is pretty minor but it seems a little weird to me that one of these methods takes a `ctx` while the other one doesn't.